test/test_metadata.py

```diff
@@ -187,3 +187,150 @@ def test_compare_deleted(self):
         m2["artist"] = "TheArtist"
         m2.delete("title")
         self.assertTrue(m1.compare(m2) < 1)
+
+    def test_metadata_mapping_init(self):
+        d = {'a': 'b', 'c': 2, 'd': ['x', 'y'], 'x': '', 'z': {'u', 'w'}}
+        deleted_tags = set('c')
+        m = Metadata(d, deleted_tags=deleted_tags, length=1234)
+        self.assertTrue('a' in m)
+        self.assertEqual(m.getraw('a'), ['b'])
+        self.assertEqual(m['d'], MULTI_VALUED_JOINER.join(d['d']))
+        self.assertNotIn('c', m)
+        self.assertNotIn('length', m)
+        self.assertIn('c', m.deleted_tags)
+        self.assertEqual(m.length, 1234)
+
+    def test_metadata_mapping_del(self):
+        m = self.metadata_d1
+        self.assertEqual(m.getraw('a'), ['b'])
+        self.assertNotIn('a', m.deleted_tags)
+
+        self.assertNotIn('x', m.deleted_tags)
+        self.assertRaises(KeyError, m.getraw, 'x')
+
+        del m['a']
+        self.assertRaises(KeyError, m.getraw, 'a')
+        self.assertIn('a', m.deleted_tags)
+
+    def test_metadata_mapping_iter(self):
+        l = set(self.metadata_d1)
+        self.assertEqual(l, {'a', 'c', 'd'})
+
+    def test_metadata_mapping_keys(self):
+        l = set(self.metadata_d1.keys())
+        self.assertEqual(l, {'a', 'c', 'd'})
+
+    def test_metadata_mapping_values(self):
+        l = set(self.metadata_d1.values())
+        self.assertEqual(l, {'b', '2', 'x; y'})
+
+    def test_metadata_mapping_len(self):
+        m = self.metadata_d1
+        self.assertEqual(len(m), 3)
+        del m['x']
+        self.assertEqual(len(m), 3)
+        del m['c']
+        self.assertEqual(len(m), 2)
+
+    def _check_mapping_update(self, m):
+        self.assertEqual(m['a'], 'b')
+        self.assertEqual(m['c'], '3')
+        self.assertEqual(m.getraw('d'), ['u', 'w'])
+        self.assertEqual(m['x'], '')
+        self.assertIn('x', m.deleted_tags)
+
+    def test_metadata_mapping_update(self):
+        # update from Metadata
+        m = self.metadata_d2
+        m2 = self.metadata_d3
+
+        del m2['x']
+        m.update(m2)
+        self._check_mapping_update(m)
+
+    def test_metadata_mapping_update_dict(self):
+        # update from dict
+        m = self.metadata_d2
+
+        d2 = {'c': 3, 'd': ['u', 'w'], 'x': ''}
+
+        m.update(d2)
+        self._check_mapping_update(m)
+
+    def test_metadata_mapping_update_tuple(self):
+        # update from tuple
+        m = self.metadata_d2
+
+        d2 = (('c', 3), ('d', ['u', 'w']), ('x', ''))
+
+        m.update(d2)
+        self._check_mapping_update(m)
+
+    def test_metadata_mapping_update_dictlike(self):
+        # update from kwargs
+        m = self.metadata_d2
+
+        m.update(c=3, d=['u', 'w'], x='')
+        self._check_mapping_update(m)
+
+    def test_metadata_mapping_update_noparam(self):
+        # update without parameter
+        m = self.metadata_d2
+
+        m.update()
```

Comment:
Actually this isn't really great, as it differs from `dict` behavior:

```
>>> d = dict()
>>> dict.update()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: descriptor 'update' of 'dict' object needs an argument
```

@phw do you think we should raise a TypeError too ?

Reply:
@phw: i did it in https://github.com/metabrainz/picard/pull/1137/commits/f6e3aedf7a9fa6cd7134fd79e18e974ead429e19
It makes more sense